rmw_connext_shared_cpp/src/node_names.cpp

```diff
@@ -67,15 +71,29 @@ get_node_names(
     return RMW_RET_BAD_ALLOC;
   }
 
-
   for (auto i = 1; i < length; ++i) {
     DDS::ParticipantBuiltinTopicData pbtd;
     auto dds_ret = participant->get_discovered_participant_data(pbtd, handles[i - 1]);
-    const char * name = pbtd.participant_name.name;
-    if (!name || dds_ret != DDS_RETCODE_OK) {
-      name = "(no name)";
+    std::string name;
+    if (DDS_RETCODE_OK == dds_ret) {
+      auto data = static_cast<unsigned char *>(pbtd.user_data.value.get_contiguous_buffer());
+      std::vector<uint8_t> kv(data, data + pbtd.user_data.value.length());
+      auto map = rmw::impl::cpp::parse_key_value(kv);
+      auto found = map.find("name");
+      if (found != map.end()) {
+        name = std::string(found->second.begin(), found->second.end());
+      }
+      if (name.empty()) {
+        // use participant name if no name was found in the user data
+        name = pbtd.participant_name.name;
+      }
+    }
+    if (name.empty()) {
+      // ignore discovered participants without a name
+      node_names->data[i] = nullptr;
+      continue;
```

Comment:
I don't think `continue`ing here is a good idea, higher layers are probably assuming that the `node_names` map will return valid strings (nothing in the `rcl_get_node_names` documentation suggests otherwise). FWICT `rclpy`, for example, is assuming null-terminated strings are stored at each index: https://github.com/ros2/rclpy/blob/fbd80a27342d56b32047ecb7f2ad2e44e9d027b3/rclpy/src/rclpy/_rclpy.c#L2323

Reply:
The line you refenced should be fine if `NULL` is passed in as far as I understand the API docs:

> If u is NULL, this function behaves like PyUnicode_FromUnicode() with the buffer set to NULL. This usage is deprecated in favor of PyUnicode_New().

Reply:
Looks like you're referencing the documentation for `PyUnicode_FromStringAndSize`, `PyUnicode_FromString` [says](https://docs.python.org/3/c-api/unicode.html):

> Create a Unicode object from a UTF-8 encoded null-terminated char buffer u.

The error (when it doesn't segfault) is: `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf0 in position 0: invalid continuation byte`

Reply:
You are right, I stopped at the first match and didn't check the exact function name :disappointed: